source/MaterialXShaderGen/HwShader.cpp

```diff
@@ -12,6 +15,26 @@ HwShader::HwShader(const string& name)
     // Create default uniform blocks for vertex stage
     createUniformBlock(VERTEX_STAGE, PRIVATE_UNIFORMS, "prv");
     createUniformBlock(VERTEX_STAGE, PUBLIC_UNIFORMS, "pub");
+
+    // Create light data uniform block with required field 'type'
+    createUniformBlock(PIXEL_STAGE, LIGHT_DATA_BLOCK, "u_lightData[MAX_LIGHT_SOURCES]");
+    createUniform(PIXEL_STAGE, LIGHT_DATA_BLOCK, DataType::INTEGER, "type");
```

Comment:
This block will hold all uniform data required by light shaders. Light shader implementations will dynamically create variables needed in this block, in their createVariables method.

The result is a single struct with data for all bound light types. And an array of this struct is supplied from application, with an entry for each active light source. In generated shader code:

```
struct LightData
{
    int type;
    vec3 position;
    vec3 direction;
    vec3 color;
    float intensity;
    float decayRate;
};

uniform LightData u_lightData[MAX_LIGHT_SOURCES];
```

Reply:
This looks like a good start. I didn't count, but it could be you need to pad the structure to have 16 byte alignment. I'm not to sure if newer layouts bypass this requirement.
e.g. https://www.opengl.org/discussion_boards/showthread.php/198866-Sending-an-array-of-structs-to-shader-via-an-Uniform-Buffer-Object

Reply:
Good point, if that is a requirement I can add that in.

Reply:
Note that the LightData struct is built up dynamically depending on the light shader types that has been bound/set on the shader generator. So for the general case an application cannot assume it always has the above members, it needs to read the corresponding VariableBlock on the returned HwShader to see which data is needed by the light shaders being used.

In practice though, for a given application you can choose to always have a fixed set of supported light types enabled, which will make the LightData struct be the same in all situations.

Reply:
I will leave padding out for now and we can add that in, if needed, when light binding support is added to the viewer.